Implemented HoC ifPropsVerify and IfPostTypeSupports.

jorgefilipecosta opened the pull request:
ifPropsVerify HoC was implemented, it will be useful in a considerable amount of places. IfPostTypeSupports HoC is a specialization of ifPropsVerify The IfPostTypeSupports replaces PostTypeSupportCheck component, using an HoC makes things easier to test as we don't need to declare more than one component per file as happened in PageAttributesOrder.
A replacement of IfPostTypeSupports was added here as a sample. Provided the new HoC's are approved and merged, all PostTypeSupportCheck and (similar components) will be replaced with HoC's.

Documentation by @aduth:
**Example Before:**

```jsx
function MyFeaturedImage() {
	return (
		<PostTypeSupportCheck supportKeys="thumbnail">
			<img />
		</PostTypeSupportCheck>
	);
}

export default connect( /* ... */ )( MyFeaturedImage );
```

**Example After:**

```jsx
function MyFeaturedImage() {
	return <img />;
}

export default compose( [
	ifPostTypeSupports( 'thumbnail' ),
	connect( /* ... */ ),
] )( MyFeaturedImage );
```

__Implementation notes:__

The thinking behind the naming convention is that we could create higher-order component-libraries with the following expectations:

Name|Behavior
---|---
`ifSomeCondition`|Renders component only if condition applies
`withSomeProps`|Renders component with additional props

__Testing instructions:__

Verify that there are no regressions in the behavior of post-type-conditional elements:

- Discussion (`comments` or `trackbacks` supports)
- Featured image (`thumbnail` support)
- Excerpt (`excerpt` support)
- Post format (`post-formats` support)
- Author (`author` support)
- Order (`page-attributes` support)
- Last Revision (`revisions` support)

aduth committed:
Editor: Replace PostTypeSupportCheck with ifPostTypeSupports

gziolo commented on editor/components/if-post-type-supports/index.js:
```diff
@@ -0,0 +1,52 @@
+ * @param   {string|Array} supportKeys An array or string The component is rendered
+ *                                     if post type supports at least one of the supports keys passed.
+ * @returns {Component}                Wrapped component.
+ */
```
Name should start with lowercase to match other HOCs. Why not start the name with ‘with’ instead of ‘if’?

gziolo commented:
I expected this HOC can be applied in more than one place. I like the implementation a lot. I’m only worried tgat it temporiraly adds so many lines of code :)

jorgefilipecosta committed:
Implemented HoC withConditionalRender and IfPostTypeSupports.

The IfPostTypeSupports will replace PostTypeSupportCheck, using HoC makes things easier to test as we don't need to declare more than one component per file as happened in PageAttributesOrder.

jorgefilipecosta commented:
Hi @aduth, I merged your PR https://github.com/WordPress/gutenberg/pull/4815 with this one I also changed the name of component withConditionalRender to ifPropsVerify to be inline with the naming convention for conditional render.

aduth commented on components/higher-order/if-props-verify/index.js:
```diff
@@ -0,0 +1,23 @@
+ */
+function ifPropsVerify( predicate ) {
+	return ( OriginalComponent ) => {
+		const WrappedComponent = ( props ) => predicate && predicate( props ) && <OriginalComponent { ...props } />;
```
Mixed messaging here: We don't trust the developer to pass a function as `predicate`, but we do trust them to not return `0` as the result of `predicate` (I'll leave it to you to experiment with the outcome 😄 ).

Hint: https://codepen.io/aduth/pen/dWwzrL

aduth replied:
So the alternative here would be:

```jsx
predicate && predicate( props ) ? <OriginalComponent { ...props } /> : null
```

youknowriad replied:
or `!! predicate && !! predicate( props ) ? <OriginalComponent { ...props } />` depends if you like the `!!` or not

aduth reviewed (commented):
@youknowriad As the original author of `PostTypeSupportCheck`, do you have any thoughts on the alternative pattern proposed here?

Something about the name `ifPropsVerify` doesn't sit well with me, though I don't have a good alternative (I actually liked `withConditionalRender`, but understand why it was changed based on my suggested naming convention).

youknowriad commented on editor/components/higher-order/if-post-type-supports/index.js:
```diff
@@ -0,0 +1,57 @@
+		WrappedComponent.displayName = getWrapperDisplayName( WrappedComponent, 'ifPostTypeSupports' );
+		return WrappedComponent;
+	};
+}
```
I think I 'd prefer a renderProp where you pass the result of the check because we have use-case where we want to do `if else`.

```js
<IfPostTypeSupports
  feature="comments" 
  renderIfTrue={ () => <div>Awesome</div> }
  renderIfFalse={ () => <div>Oh No!</div> }
/>
```

I think the higher order component could still be exposed as a special case though.

aduth replied:
Hm, I agree this would be difficult to implement with higher-order components.

The prop naming doesn't jive with me, and I'd be more inclined toward a function-as-child alternative, but for no particularly strong reasons:

```jsx
<IfPostTypeSupports feature="comments">
	{ ( isSupported ) => (
		<div>{ isSupported ? 'Awesome' : 'Oh No!' }</div>
	) }
</IfPostTypeSupports>
```

_(Though in this trivial example it does help for reducing code duplication on shared wrappers, although in the real world this would probably be wrapping the `<IfPostTypeSupports />` anyways)_

jorgefilipecosta replied:
Hi @youknowriad, @aduth, I thought a little and I think we can keep the simplicity of HoC and still allow rendering an else condition.
If we use the HoC like:
ifPostTypeSupports( 'thumbnail' ),

The HoC does not render the component, used in most of the cases.

But if we use the HoC like:
ifPostTypeSupports( 'thumbnail', 'isThumbnailSupported' ),
The component is rendered with the additional prop isThumbnailSupported passed so the user of the HoC can implement its if/else logic.

Let me know your thoughts on this approach.

aduth replied:
Yeah, that works pretty well. Would it be the standard usage? Or would there be a default inferred usage that if the second argument isn't specified, the component doesn't render at all if the condition isn't met?†

_† And if so, is this potentially confusing in its inconsistency?_

jorgefilipecosta replied:
My idea was that the components don't render at all if the second argument isn't specified. It may be a not easy to guess behavior but if we document it I think users of the component will get familiar with it.
